frontend/test/pytest/test_autograph.py

```diff
@@ -1343,6 +1343,153 @@ def f1():
         with pytest.raises(RuntimeError):
             qjit(autograph=True)(f1)()
 
+    def test_uninitialized_variables(self, monkeypatch):
+        """Verify errors for (potentially) uninitialized loop variables."""
+        monkeypatch.setattr("catalyst.autograph_strict_conversion", True)
+
+        def f(pred: bool):
+            while pred:
+                x = 3
+
+            return x
+
+        with pytest.raises(AutoGraphError, match="'x' is potentially uninitialized"):
+            qjit(autograph=True)(f)
+
+    def test_init_with_invalid_jax_type(self, monkeypatch):
+        """Test loop carried values initialized with an invalid JAX type."""
+        monkeypatch.setattr("catalyst.autograph_strict_conversion", True)
+
+        def f(pred: bool):
+            x = ""
+
+            while pred:
+                x = 3
+
+            return x
+
+        with pytest.raises(AutoGraphError, match="'x' was initialized with type <class 'str'>"):
+            qjit(autograph=True)(f)
+
+    def test_init_with_mismatched_type(self, monkeypatch):
+        """Test loop carried values initialized with a mismatched type compared to the values used
+        inside the loop."""
+        monkeypatch.setattr("catalyst.autograph_strict_conversion", True)
+
+        def f(pred: bool):
+            x = 0.0
+
+            while pred:
+                x = 3
+
```

Comment:
Is type promotion supported here?

I suppose the following code should pass:

``` python

        def f(pred: bool):
            x = 1

            while pred:
                x = 3.2

            return x
```

Reply:
Hmm good question, I believe it is not supported. @grwlf recently added type promotion for conditionals, but not for loops.

Autograph will raise an error because of the mismatched types, and then a fallback to Python will trigger here, which will actually fail because `pred` is a tracer 🤔

I wonder if it best to actually ensure the original exception is raised, rather than falling back to Python. But distinguishing these cases is increasingly adding complexity.

Reply:
If this will end up failing either from Python or AG, why not raising the original exception from AG to reduce the boilerplate code? I assume you already have given this some thought, but I want to learn more about its inner working here.

Reply:
> If this will end up failing either from Python or AG, why not raising the original exception from AG to reduce the boilerplate code?

I totally agree, we should raise our AG-type exception instead of the one the fallback is raising, but doing so will not reduce boilerplate code, rather it will add more code (as we can't get rid of the fallback mechanism, other scenarios depend on it).
Rather, the reason to raise the original exception is actually that the user will be better able to fix their mistake, since here they clearly intended for the conversion to happen (due to the dynamic `pred`). The mistake they can fix is the incorrectly initialized variable.

I haven't given this a lot of thought for the while loop, since @grwlf is the one who implemented it. I think the best thing for now is adding this as a todo to allow the bug addressed in this PR to be merged. What do you think?

Thanks for bringing it up though, we definitely need to address it!

Reply:
I turned this comment into issue #354